### PR TITLE
Add support for date display with local preference, refs 1545, 2509

### PR DESCRIPTION
**Pull request by mwjames**
This PR is made in reference to: #1545, #2509

This PR addresses or contains:

- User can enable the preference (`smw-prefs-general-options-time-correction`) to display local corrected dates in `Special:Browse` and the property page
- Corrected dates will be hinted with a `ᴸ` as in `05:41:20, 31 May 2017 ᴸ`
- Query output formatter can use `LOCL#TO` as option to adapt a local preference
- Use `ParserOptionsRegister` (MW 1.30+) to register an extra options key (`localTime`) with the `ParserCache` to ensure that users with different settings can switch the view without retaining a cached (== old settings) page view

This PR includes:
- [x] Tests (unit/integration)
- [x] CI build passed

### CODE REVIEW

File: src/DataValues/Time/LocalTime.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	 * @param DateTime $dateTime
+	 * @param User|null $user
+	 *
+	 * @return DateTime
+	 */
+	public static function adjustTimeByPreference( DateTime $dateTime, User $user = null ) {
+
+		$tz = false;
+
+		if ( $user === null  ) {
```
Comment (gesinn-it) [inline src/DataValues/Time/LocalTime.php]: shouldn't this be `if ( $user )`? We did a test with the fix, and now it works instead of using the defaults.

**Comment by mwjames** [comment]
> +		if ( $user === null  ) {
>
> shouldn't this be `if ( $user )`? We did a test with the fix, and now it
> works instead of using the defaults.

Probably what I wanted here was something a long the line of:

```
if ( $user === null  ) {
 $user = $GLOBALS['wgUser'];
}

$tz = $user instanceof User ? $user->getOption( 'timecorrection' ) : false;
```

It is used as fallback to ensure that the method can be tested with an
injected `User` object, if it is tested, which obviously is not.

Also instead of calling `LocalTime` directly, call
`Localizer::getInstance()->getLocalTime( $dateTime )` to avoid
`TimeValueFormatter` to depend on a MW object state (e.g wgUser) with:

 - Localizer::getLocalTime calling out to
`LocalTime::getLocalizedTime( $dateTime, $GLOBALS['wgUser'] );` and
- Move `LocalTime` to the \SMW\MediaWiki namespace as it is a pure
MediaWiki thing that relies entirely on MW settings and object state.

**Comment by gesinn-it** [comment]
@mwjames , okay, we'll try to help with adding tests.

Is `https://github.com/SemanticMediaWiki/SemanticMediaWiki/tree/local-time/tests/phpunit/Unit/DataValues/Time` the right place to add the UTs for _LocalTime_?

**Comment by mwjames** [comment]
> `https://github.com/SemanticMediaWiki/SemanticMediaWiki/tree/local-time/tests/phpunit/Unit/DataValues/Time`
> the right place to add the UTs for _LocalTime_?

If we neglect my last comment, then yes.

**Comment by mwjames** [comment]
https://www.semantic-mediawiki.org/wiki/Local_time

**Comment by mwjames** [comment]
In order to show local times on special pages (SMW), the option has to be enabled using the user preference.

image

Dates with a modified local time are marked with a `ᴸ` where `ᴸ` will reveal the ISO date on a mouseover event. 

image

**Pull request merged**